### PR TITLE
Add links to company websites for breach resolution

### PR DESCRIPTION
**Pull request by flozia**
# References: 
Jira: [MNTOR-1504](https://mozilla-hub.atlassian.net/browse/MNTOR-1504)

# Description

Adds a link to the breach resolution when passwords or security questions were involved.

# Screenshot

MNTOR-1504

# How to test

1. Login with an account or add an email address that was exposed in a breach.
2. Visit `/user/breaches`
3. If the breach includes passwords or security questions there should be a link to the company website in the associated resolution step.

# Checklist (Definition of Done)
- [x] Localization strings (if needed) have been added.
- [x] Commits in this PR are minimal and [have descriptive commit messages](https://chris.beams.io/posts/git-commit/).
- [x] I've added or updated the relevant sections in readme and/or code comments
- [x] I've added a unit test to test for potential regressions of this bug.
- [x] Product Owner accepted the User Story (demo of functionality completed) or waived the privilege.
- [x] All acceptance criteria are met.
- [x] Jira ticket has been updated (if needed) to match changes made during the development process.
- [x] Jira ticket has been updated (if needed) with suggestions for QA when this PR is deployed to stage.

### CODE REVIEW

File: src/utils/breach-resolution.js [hunk 1]
```diff
@@ -114,6 +114,7 @@ function appendBreachResolutionChecklist (userBreachData, options = {}) {
       const dataClasses = b.DataClasses
       const args = {
         companyName: b.Name,
+        breachedCompanyLink: `<a href="https://${b.Domain}" target="_blank">${b.Domain}</a>`,
```
Comment (Vinnl) [inline src/utils/breach-resolution.js:117]: Issue, blocking: sometimes `.Domain` is an empty string. We'll probably want to either keep the previous string for those cases, or hide the entire suggestion in those cases.

**Review by Vinnl — COMMENTED** [review]
Left on blocking comment, unfortunately, though probably fairly easy to resolve :)

**Comment by pdehaan** [comment]
Not sure if I mentioned this in another thread, but there is a small-ish risk of linking to external sites. A lot of the domains via HIBP are no longer resolving.

I looked at one point and think my scripts were guessing about 25% of the outbound domains were invalid.
I had a few minutes so I ended up rebuilding my tool: https://github.com/pdehaan/blurts-https-stats which conveniently pipes the report out to README.md if you just want a quick scroll. Or, the raw "alive"-vs-"dead" JSON can be found in https://github.com/pdehaan/blurts-https-stats/blob/main/checker.json.

Some very rough stats, if you like grepping JSON files:
- `"link":` occurs 624&times; (which accounts for empty domains or duplicated domains)
- `"status": "dead"` occurs 198&times; (31.7%)
- `"status": "alive"` occurs 426&times; (68.2%)

_BUT_ the misleading thing is that a lot of those broken 31% reported dead links still work, but the npm module is just reporting bad HTTPS certificates as broken (or getting caught by Cloudflare captchas).

---

**PROTIP:** If you want to play a fun game, we could WHOIS all the broken domains and see which domains are available to register, then pick them all up and redirect them to https://monitor.firefox.com.

File: locales/en/breaches.ftl [hunk 2]
```diff
@@ -136,7 +136,7 @@ breach-checklist-phone-header-2 = Protect your phone number with a masking servi
 ## Prompts the user for changes when there is a breach detected of security questions
 
 # { $breachedCompanyLink } will link to the website of the company where the breach occurred
-breach-checklist-sq-header-2 = Update your security questions on the company’s website.
+breach-checklist-sq-header-3 = Update your security questions on { $breachedCompanyLink }.
```
Comment (flodolo) [inline locales/en/breaches.ftl:139]: I don't understand these changes. I thought we agreed a couple of weeks ago that we don't want links?
https://github.com/mozilla/blurts-server/pull/2900#discussion_r1135799859

**Review by flodolo — CHANGES_REQUESTED** [review]
Requesting changes to make sure the rationale is clarified.

File: locales/en/breaches.ftl (hunk 2 above)
Comment (flozia) [reply locales/en/breaches.ftl:139]: Yes, there was some back and forth. Ultimately PM decided to add those links and [MNTOR-1504](https://mozilla-hub.atlassian.net/browse/MNTOR-1504) was created.

Comment (flozia) [reply locales/en/breaches.ftl:139]: cc @EMMLynch

**Comment by flozia** [comment]
> Not sure if I mentioned this in another thread, but there is a small-ish risk of linking to external sites. A lot of the domains via HIBP are no longer resolving.
> 
> I looked at one point and think my scripts were guessing about 25% of the outbound domains were invalid. I had a few minutes so I ended up rebuilding my tool: https://github.com/pdehaan/blurts-https-stats which conveniently pipes the report out to README.md if you just want a quick scroll. Or, the raw "alive"-vs-"dead" JSON can be found in https://github.com/pdehaan/blurts-https-stats/blob/main/checker.json.
> 
> Some very rough stats, if you like grepping JSON files:
> 
> * `"link":` occurs 624× (which accounts for empty domains or duplicated domains)
> * `"status": "dead"` occurs 198× (31.7%)
> * `"status": "alive"` occurs 426× (68.2%)
> 
> _BUT_ the misleading thing is that a lot of those broken 31% reported dead links still work, but the npm module is just reporting bad HTTPS certificates as broken (or getting caught by Cloudflare captchas).
> 
> **PROTIP:** If you want to play a fun game, we could WHOIS all the broken domains and see which domains are available to register, then pick them all up and redirect them to https://monitor.firefox.com.

That’s interesting. Thank you so much for the tool @pdehaan  and for listing the stats! As @flodolo mentioned there had been some back and forth on if we would like to show those links or not. We communicated the risks with PM but I’m raising this again to make sure we are aligned.

**Comment by pdehaan** [comment]
@flozia To be clear, the broken links are already on the site on the details page, ie: https://monitor.firefox.com/breach-details/LeakedReality
If I click on the domain near the top I see a _"Hmm. We’re having trouble finding that site."_ error page in Firefox Nightly. A WHOIS search shows the domain is registered, but maybe DNS just isn't set up anymore. Even curl shows an error trying to inspect the page:
```sh
curl -fL https://leakedreality.com # HTTPS
curl: (6) Could not resolve host: leakedreality.com

curl -fL http://leakedreality.com # try HTTP
curl: (6) Could not resolve host: leakedreality.com

echo $? # 6
```

&hellip; but I can see how adding a new [broken] CTA link would be frustrating and confusing to end users.

---

It took a few clicks, but https://monitor.firefox.com/breach-details/Abandonia2022 is another example. Clicking the domain link on the details page takes me to a "Unable to connect. An error occurred during a connection to abandonia.com." error page in Nightly. Changing `https://` to vanilla `http://` fixes the broken link, but probably not intuitive.

Or https://monitor.firefox.com/breach-details/GGCorp external domain link takes me to a scary error page (presumably because of a bad/invalid HTTPS cert):
> **Warning: Potential Security Risk Ahead**
> Nightly detected a potential security threat and did not continue to ggcorp.me. If you visit this site, attackers could try to steal information like your passwords, emails, or credit card details.

**Comment by mansaj** [comment]
@pdehaan Looking at the JSON generated by the script..  I noticed that lots of the sites marked as “dead” / 403s are accessible? Are these false positives?

**Comment by pdehaan** [comment]
> @mansaj https://github.com/mozilla/blurts-server/pull/2961#issuecomment-1494578082 "I noticed that lots of the sites marked as “dead” / 403s are accessible? Are these false positives?"

I think the [**link-check**](https://www.npmjs.com/package/link-check) module I used might be considering certificate errors to be "dead" links? So there were sites that were reported as dead, but clicking them showed me the website but there was some console logging and/or certificate issues that most people wouldn't really notice. Per https://github.com/tcort/link-check/issues/47#issuecomment-847764254 there might also be some 403 issues if sites are behind Cloudflare or maybe other proxy services as well.

Per link-check README:
> A link is said to be 'alive' if an HTTP HEAD or HTTP GET for the given URL eventually ends in a `200 OK` response. To minimize bandwidth, an HTTP HEAD is performed. If that fails (e.g. with a `405 Method Not Allowed`), an HTTP GET is performed. Redirects are followed.

GitHub search showed this as one of the very few results for "dead" in their codebase: https://github.com/tcort/link-check/blob/430027f6be03b904db508042945bf2a75472d330/lib/LinkCheckResult.js#L10

I can't think of another great solution. I tried using `curl` or native `fetch` instead but neither were great, both were super slow (these scripts take 10-20 minutes to scrape 640 domains w/ 10s timeouts). Another option might be using something like Playwright and trying to click the link from the `/breach-details/*` page and then taking a screenshot if it's a non-2xx response. If the screenshot shows _something_ maybe we don't care. But if we get no response/bytes/pixels, maybe we consider that a bad link. Although I imagine that's still prone to Cloudflare sandboxes and other gotchas.

**Review by flodolo — CHANGES_REQUESTED** [review]

File: src/app-constants.js [hunk 3]
```diff
@@ -73,4 +79,40 @@ optionalEnvVars.forEach(key => {
   if (value) AppConstants[key] = value
 })
 
+// Create HIBP breach link blocklist
+const linkListData = readFileSync(path.join(
```
Comment (flozia) [inline src/app-constants.js:83]: I’m not sure what we should do for the first iteration before automating things:
1. Add the JSON in this repository as is
2. Add `blurts-https-stats` as a submodule
3. Set `HIBP_BREACH_LINK_BLOCKLIST` directly in `.env`

Currently I’m leaning towards the third options. What do you think @Vinnl, @mansaj and @pdehaan?

File: locales/en/breaches.ftl [hunk 4]
```diff
@@ -68,7 +68,13 @@ breach-checklist-link-mozilla-vpn = { -brand-mozilla-vpn }
 ## Prompts the user for changes when there is a breach detected of password
 
 # { $breachedCompanyLink } will link to the website of the company where the breach occurred
-breach-checklist-pw-header-2 = Go to the company’s website to change your password and enable two-factor authentication (2FA).
+breach-checklist-pw-header-3 =
+  {
+    $breachedCompanyLink ->
+      [empty] Go to the company’s website to change your password and enable two-factor authentication (2FA).
```
[the remark this answers is not in the document]
Comment (flozia) [reply locales/en/breaches.ftl]: Makes sense — separated the strings in [aa00a5a](https://github.com/mozilla/blurts-server/pull/2961/commits/aa00a5a275638c145aecd989b0dc7678b222eca0).

File: locales/en/breaches.ftl [hunk 5]
```diff
@@ -67,8 +67,10 @@ breach-checklist-link-mozilla-vpn = { -brand-mozilla-vpn }
 
 ## Prompts the user for changes when there is a breach detected of password
 
-# { $breachedCompanyLink } will link to the website of the company where the breach occurred
 breach-checklist-pw-header-2 = Go to the company’s website to change your password and enable two-factor authentication (2FA).
+# { $breachedCompanyLink } will link to the website of the company where the breach occurred
+breach-checklist-pw-header-3 = Go to { $breachedCompanyLink } to change your password and enable two-factor authentication (2FA).
```
Comment (flodolo) [inline locales/en/breaches.ftl:72]:
```suggestion
breach-checklist-pw-header-link = Go to { $breachedCompanyLink } to change your password and enable two-factor authentication (2FA).
```

Let's not use 3, since this is not versioned string anymore.

File: locales/en/breaches.ftl [hunk 6]
```diff
@@ -135,8 +137,10 @@ breach-checklist-phone-header-2 = Protect your phone number with a masking servi
 
 ## Prompts the user for changes when there is a breach detected of security questions
 
-# { $breachedCompanyLink } will link to the website of the company where the breach occurred
 breach-checklist-sq-header-2 = Update your security questions on the company’s website.
+# { $breachedCompanyLink } will link to the website of the company where the breach occurred
+breach-checklist-sq-header-3 = Update your security questions on { $breachedCompanyLink }.
```
Comment (flodolo) [inline locales/en/breaches.ftl:142]:
```suggestion
breach-checklist-sq-header-link = Update your security questions on { $breachedCompanyLink }.
```

Same here, let's use a better ID.

File: src/app-constants.js (hunk 3 above)
Comment (Vinnl) [reply src/app-constants.js:83]: If 3. is no problem for Brandon, that sounds good to me. Otherwise 1. also sounds absolutely fine to me.

Comment (Vinnl) [reply src/app-constants.js:83]: @pdehaan It's not super easy by any chance to also output how long ago every breach was whose website did not resolve? That could give us an indication of how soon new breaches are likely to stop resolving.

Comment (pdehaan) [reply src/app-constants.js:83]: > It's not super easy by any chance to also output how long ago every breach was whose website did not resolve? That could give us an indication of how soon new breaches are likely to stop resolving.

I'd have to think it over. I think my problem is that I'm generally looping over unique `domains`, not breaches. So there isn't always an easy way to find an exact breach by domain (although probably that's just more of an edge case for the [15?] domains that have more than 1 breach).

Comment (pdehaan) [reply src/app-constants.js:83]: @Vinnl https://github.com/pdehaan/blurts-domains-playwright
Now with tables, so it looks something more like

| STATUS | DOMAIN | BREACH DATE | URL
|:----:|:----|:----:|:----
200/OK | adobe.com | 10/3/2013 | https://www.adobe.com/

Comment (Vinnl) [reply src/app-constants.js:83]: Awesome! So manually checking the errored ones from 2020 and later, only the shockgore and tamodo domains no longer resolved for me, both from 2020. So I expect we're "safe" for a while if using a "stale" list, and to be extra sure, we could also use a cut-off date of two or one years for when to link to a breached company.

Comment (pdehaan) [reply src/app-constants.js:83]: But one or two years from _when_? BreachDate? AddedDate? ModifiedDate?
At one point I had a handy calculator that did displayed the number of days between the BreachDate and AddedDate and saw that some breaches weren't added to the HIBP site until like 3-5 years after the BreachDate.

Comment (pdehaan) [reply src/app-constants.js:83]: https://github.com/pdehaan/blurts-relative-breach-date
TLDR:

- 668 breaches
- AVG DATE DIFF: 472.73 days (1.3 years)
- 174 (26%) have a diff measured in "days" (presumably less than 30 days)
- 243 (36%) have a diff measured in "mos" (greater than 30 days, but less than 12.0 months)
- 248 (37%) have a diff measured in "yrs" (greater than 12 months)

(i'm missing 3 breaches, so there must be a month or year singular that my Ctrl+F isn't picking up -- and I'm too lazy to script/grep it)

**Comment by flozia** [comment]
Thanks for your input, everyone! After this PR was on hold while aligning with PM and UX, I’m moving this out of “draft” mode.

1. We decided to compile a first blocklist from @pdehaan’s https://github.com/pdehaan/blurts-domains-playwright/blob/main/stats-https.json. We will use the `env` variable `HIBP_BREACH_LINK_BLOCKLIST` which will include domains that do not resolve with `200`.
2. @flodolo The strings were updated and include `b` tags and a marker `breached-company-link` that can be replaced by a link — or stripped if it is one we do not want to show.

**Comment by flodolo** [comment]
> 2\. @flodolo The strings were updated and include `b` tags and a marker `breached-company-link` that can be replaced by a link — or stripped if it is one we do not want to show.

When you say stripped, you mean that the markup is removed, but the text remains? i.e. it's become inactive text

**Review by flodolo — APPROVED** [review]
(never mind my previous comment, the code is clear enough)

File: src/utils/breachResolution.js [hunk 7]
```diff
@@ -141,32 +147,72 @@ function getResolutionRecsPerBreach (dataTypes, args, options = {}) {
...
+ * @param {string} body for the fluent body string
+ * @param {object} args
+ * @returns {string} body string
+ */
+function getBodyMessage (body, args) {
+  const { resolutionType, stringArgs } = args
+
+  switch (resolutionType) {
+    case BreachDataTypes.Passwords:
+    case BreachDataTypes.SecurityQuestions: {
```
Comment (Vinnl) [inline src/utils/breachResolution.js:189]: Personal preference, so feel free to ignore, but I'd keep the code simple here and just replace `<breached-company-link>` in every string, rather than just the passwords and security questions resolutions.

File: src/utils/breachResolution.test.js [hunk 8]
```diff
@@ -147,13 +149,22 @@ test('appendBreachResolutionChecklist: data class with a resolution referring to
     unverifiedEmails: []
   }
   appendBreachResolutionChecklist(userBreachData)
-  // There should only be a resolution for `BreachDataTypes.Phone`, as
-  // `BreachDataTypes.Passwords` and `BreachDataTypes.SecurityQuestions` refer
-  // to the breached company's domain, which we don't know:
+  // There should be a resolution for `BreachDataTypes.Phone`,
+  // `BreachDataTypes.Passwords` and `BreachDataTypes.SecurityQuestions`.
+  // The last two should fallback to a more generic header string that does not
+  // include the breached company's domain, which we don't know:
```
Comment (Vinnl) [inline src/utils/breachResolution.test.js:155]: I don't know how easy it is to mock AppConstants, but if it's easy (but only then - you've been working on this PR for long enough), maybe a test for the blocklist would be a good addition?

File: tsconfig.json [hunk 9]
```diff
@@ -36,9 +36,9 @@
     // "useDefineForClassFields": true,                  /* Emit ECMAScript-standard-compliant class fields. */
     "moduleDetection": "force", /* Control what method is used to detect module-format JS files. */
     /* Modules */
-    "module": "commonjs", /* Specify what module code is generated. */
+    "module": "ES2022", /* Specify what module code is generated. */
     "rootDir": "./src/", /* Specify the root folder within your source files. */
-    // "moduleResolution": "node10",                     /* Specify how TypeScript looks up a file from a given module specifier. */
+    "moduleResolution": "node",                          /* Specify how TypeScript looks up a file from a given module specifier. */
```
Comment (Vinnl) [inline tsconfig.json:41]: Hoowee this looks like a potentially far-reaching change that's probably outside of the scope of this PR - an accidental commit?

**Review by Vinnl — APPROVED** [review]
Only easy fixes or things that can be ignored, I think :)

File: tsconfig.json (hunk 9 above)
Comment (flozia) [reply tsconfig.json:41]: Reverted.

File: src/utils/breachResolution.js [hunk 10]
```diff
@@ -34,8 +35,8 @@ const BreachDataTypes = {
 const breachResolutionDataTypes = {
   [BreachDataTypes.Passwords]: {
     priority: 1,
-    header: 'breach-checklist-pw-header-2',
-    body: 'breach-checklist-pw-body-2'
```
[the remark this answers is not in the document]
Comment (flozia) [reply src/utils/breachResolution.js:38]: Got it — thanks!

File: src/utils/breachResolution.js [hunk 11]
```diff
@@ -141,32 +147,72 @@ function getResolutionRecsPerBreach (dataTypes, args, options = {}) {
   const filteredBreachRecs = {}
 
   // filter breachResolutionDataTypes based on relevant data types passed in
-  for (const [key, value] of Object.entries(breachResolutionDataTypes)) {
+  for (const resolution of Object.entries(breachResolutionDataTypes)) {
+    const [key, value] = resolution
     if (
       dataTypes.includes(key) &&
-      // Hide the security question or password resolution if we can't link to the breached site:
```
[the remark this answers is not in the document]
Comment (flozia) [reply src/utils/breachResolution.js:147]: Removed in [54ce192](https://github.com/mozilla/blurts-server/pull/2961/commits/54ce192f397b22d40bda6454c68087466f4009e8).

File: .env-dist [hunk 12]
```diff
@@ -58,6 +58,8 @@ HIBP_THROTTLE_DELAY=2000
 HIBP_THROTTLE_MAX_TRIES=5
 # Authorization token for HIBP to present to /hibp/notify endpoint
 HIBP_NOTIFY_TOKEN=unsafe-default-token-for-dev
+# Domains we prefer to not link to
+HIBP_BREACH_DOMAIN_BLOCKLIST=a-blocked-domain.com,another-blocked-domain.org
```
Comment (pdehaan) [inline .env-dist:62]: Is there a limit how long an ENV var can be? 
It feels like this could be REALLY long if we end up blocking over 20-30 domains.

File: src/utils/breachResolution.js [hunk 13]
```diff
@@ -109,19 +110,24 @@ const breachResolutionDataTypes = {
...
+
   for (const { breaches } of verifiedEmails) {
     breaches.forEach(b => {
       const dataClasses = b.DataClasses
+      const blockList = (AppConstants.HIBP_BREACH_DOMAIN_BLOCKLIST ?? '').split(',')
+      const showLink = b.Domain && !blockList.includes(b.Domain)
+
       const args = {
         companyName: b.Name,
+        breachedCompanyLink: showLink ? `https://${b.Domain}` : '',
```
Comment (pdehaan) [inline src/utils/breachResolution.js:122]: aside: in my personal testing, I think `http://` had better results than `https://` (somewhere between 5-10% more 2xx/3xx results).

File: src/utils/breachResolution.test.js (hunk 8 above)
Comment (flozia) [reply src/utils/breachResolution.test.js:155]: Good point! I’ll address this in a follow-up in order to not block this PR.

File: src/utils/breachResolution.js (hunk 7 above)
Comment (flozia) [reply src/utils/breachResolution.js]: No hard preference from my side: I thought being a bit more specific here was OK since we were cautious about how and where we link out to. Less complex is also a good thing, though: [d574bee](https://github.com/mozilla/blurts-server/pull/2961/commits/d574bee881fa1d55364509f8e1735a34596204dd).

File: src/utils/breachResolution.js (hunk 13 above)
Comment (flozia) [reply src/utils/breachResolution.js:122]: Interesting, thanks for the note. With us trying to be cautious where we link out to I think I’d feel more comfortable linking out to `https://`.

File: .env-dist (hunk 12 above)
Comment (flozia) [reply .env-dist:62]: According to SRE, there is no limit that we would hit.

Comment (pdehaan) [reply .env-dist:62]: I'm wondering if it's better as an ENV var that has to be coordinated w/ SRE, versus maybe some JSON file that lives in the repo that is a single source of truth that we can audit occasionally. (unless there are reasons to keep it as a secret/ENV). 🤷 
Although I guess I could technically recreate the blocklist locally by scraping the 650 breaches on the Monitor site and see if the outbound link is a link or not.

Comment (flozia) [reply .env-dist:62]: I’m still not sure about that as well, but one good argument for handling the list in the env is that we would be able to make adjustments without a release. Especially in the beginning, when we might need to test and audit the sites manually.

**Comment by flozia** [comment]
> (never mind my previous comment, the code is clear enough)

Thank you for your renewed review @flodolo. Unfortunately, there has been another change to the strings with two additions: A [note on using 2FA](https://github.com/mozilla/blurts-server/pull/2961/commits/e05ea1a53e46db559911407bdc1b00af0167252a#diff-fa86debcc8cf851252f6487b2b0eaf01e6572734a42de03635742ab551faeb7eR70) and the[ link to Firefox Password Manager](https://github.com/mozilla/blurts-server/pull/2961/commits/e05ea1a53e46db559911407bdc1b00af0167252a#diff-fa86debcc8cf851252f6487b2b0eaf01e6572734a42de03635742ab551faeb7eR75) — sorry to burn your cycles on these.

**Review by flodolo — APPROVED** [review]
